modepy/test/test_tools.py

```diff
@@ -596,6 +596,37 @@ def test_tensor_product_vdm_dim_by_dim(dim):
 # }}}
 
 
+# {{{ test_tensor_product_pickle
+
+def test_tensor_product_pickling():
+    import pickle
+
+    hc = mp.Hypercube(3)
+    tp_space = mp.space_for_shape(hc, 1)
+
+    file = open(b"hypercube-shape.obj", "wb")
+    pickle.dump(hc, file)
+    file.close()
```

Comment:
```suggestion
    pkl = pickle.dumps(hc)
    reloded_hc = pickle.loads(pkl)
    assert hc == reloaded_hc
```
? There's no need to dump to a file. Just a small nitpick though, so feel free to just leave it as is.

Reply:
I've made this change locally for the Hypercube, but doing the same assert check with the TensorProductSpace object throws an error. The original and reloaded objects seem to be the same. It seems the bases themselves are what is causing the assertion to fail. Is this expected?

Reply:
Hm, that's weird, can't say I have an idea of why that's failing. Can you push your WIP version and I can take a look too?

Reply:
Ah, figured it out!

* The `mp.Shape` is a dataclass and defines equality.
* The `mp.FunctionSpace` is just a standard class, so it's equality is defined based on `id(obj)`, so it will fail.

We can define `__eq__` and `__hash__` for `FunctionSpace` as well, but it's probably not needed. I updated the test to just check the `spatial_dim` and the `space_dim` instead of the whole object.

Reply:
I see, that makes sense! Sounds good, thanks for making that change